### PR TITLE
[Draft] Delta firmware updates post

### PR DESCRIPTION
**Pull request by franc0is**

### CODE REVIEW

**Comment by franc0is** [comment]
I recommend looking at commit 07af81d2b0e7da02ff5598ff857dc5ae752be9be to review the example code.

File: _drafts/ota-delta-updates.md [hunk 1]
```diff
@@ -0,0 +1,600 @@
...
+downsides worth considering:
+
+1. More complexity! Your firmware update code now needs to implement the
+   patching algorithm, which creates more opportunities for bugs.
+2. Delta updates requires more code space to implement the patching algorithm,
+   so if you are constrained already it may not be appropriate for your project.
+3. Managing delta updates creates more complexity on the backend and on the
+   release process.
+
+That last point bears elaborating a bit. Unlike all system images, each delta
```
Comment (noahp) [inline _drafts/ota-delta-updates.md:101]: Algorithms like casync, zsync are designed to have the client compute the delta, not sure if it's worth mentioning though

File: _drafts/ota-delta-updates.md [hunk 2]
```diff
@@ -0,0 +1,600 @@
...
+
+> Our OTA backend at [Memfault](https://memfault.com) supports delta updates!
+> You can read more about in our
+> [documentation](https://docs.memfault.com/docs/platform/ota/#delta-releases).
+
+### Jojodiff
+
+One of the key components of a delta update system is a binary diff and patch
+system. There are remarkably few libraries that provide this functionality. The
+excellent BSDiff[^bsdiff], and XDelta[^xdelta] both require too much memory to
```
Comment (noahp) [inline _drafts/ota-delta-updates.md:120]: These algorithms don't require loading the patched binary into memory like they say on the box, though, and can work just fine without much memory at all (source: we used both at Fitbit). You do I think have to tweak the implementations a tiny bit to handle that though.

File: _drafts/ota-delta-updates.md [hunk 3]
```diff
@@ -0,0 +1,600 @@
...
+
+The resulting `patch.bin` is 1252 bytes, down from 7908 bytes for the full app
+image. This is only a ~6x reduction in size, but in practice you will see much
+better result on larger and more complex applications.
+
+Since we're not concerned with downloading the image from the cloud for the sake
+of this blog post, we just hardcode the patch.bin inside of our Loader image
+using `objcopy` as we explained in previous posts.
+
+> Note: C23 which was just finalized will include a new `#embed` preprocessor
```
Comment (noahp) [inline _drafts/ota-delta-updates.md:287]: 🙌

File: _drafts/ota-delta-updates.md [hunk 4]
```diff
@@ -0,0 +1,600 @@
...
+
+static unsigned char source_buf[4096];
+static unsigned char target_buf[4096];
+static unsigned char patch_buf[4096];
+
+int do_janpatch() {
+    janpatch_ctx ctx = {
+        // fread/fwrite buffers for every file, minimum size is 1 byte
+        // when you run on an embedded system with block size flash, set it to the size of a block for best performance
+        { source_buf, 4096 },
```
Comment (noahp) [inline _drafts/ota-delta-updates.md:341]: Maybe sizeof? Or maybe it's more clear as explicit value

File: Gemfile [hunk 5]
```diff
@@ -21,3 +21,5 @@ gem "jekyll-target-blank", "~> 2.0"
 gem "webrick", "~> 1.7"
 
 gem "jekyll-get-json", "~> 1.0"
+
+gem "eventmachine", "~> 1.3.dev.1", :git => "https://github.com/eventmachine/eventmachine.git"
```
Comment (noahp) [inline Gemfile:25]: What was this for?

**Review by noahp — APPROVED** [review]
Amazing article, love it! Maybe drop it through grammarly to catch any of the typos? Otherwise I day :shipit:

File: _drafts/ota-delta-updates.md [hunk 6]
```diff
@@ -0,0 +1,600 @@
...
+
+{% include submit-pr.html %}
+<!-- Interrupt Keep END -->
+
+{:.no_toc}
+
+## References
+
+<!-- prettier-ignore-start -->
+[^reference_key]: [Post Title](https://example.com)
```
Comment (noahp) [inline _drafts/ota-delta-updates.md:594]: Nit remove

File: _drafts/ota-delta-updates.md (hunk 1 above)
Comment (franc0is) [reply _drafts/ota-delta-updates.md:101]: Decided not to include this

File: _drafts/ota-delta-updates.md (hunk 2 above)
Comment (franc0is) [reply _drafts/ota-delta-updates.md:120]: Added a note.

File: _drafts/ota-delta-updates.md (hunk 4 above)
Comment (franc0is) [reply _drafts/ota-delta-updates.md:341]: Added a `#define`

File: Gemfile (hunk 5 above)
Comment (franc0is) [reply Gemfile]: Reverted (but it's necessary on M1 😢)

File: _drafts/ota-delta-updates.md (hunk 6 above)
Comment (franc0is) [reply _drafts/ota-delta-updates.md:594]: 👍

**Review by tyhoff — APPROVED** [review]

**Comment by tyhoff** [comment]
If you have the time, run the post through Grammarly. If not, I'll do it before publishing it so no worries.

File: _drafts/ota-delta-updates.md [hunk 7]
```diff
@@ -0,0 +1,607 @@
...
+
+> Note: the next few sections provide an overview of delta updates, why it
+> matters, and how it works. If you'd like to skip straight to implementation,
+> click
+> [here](#delta-updates-implementation).
+
+The concept behind delta updates is intuitive. When we put together a new
+version of our firmware, we do not change every single line of code. Instead, we
+edit specific sections of the code to change the behavior of the code and
+introduce a new feature or fix a bug.
```
[the remark this answers is not in the document]
Comment (franc0is) [reply _drafts/ota-delta-updates.md:54]: While you're technically correct I don't think this is too important to specify

File: _drafts/ota-delta-updates.md [hunk 8]
```diff
@@ -0,0 +1,607 @@
...
+2. Delta updates requires more code space to implement the patching algorithm,
+   so if you are constrained already it may not be appropriate for your project.
+3. Managing delta updates creates more complexity on the backend and on the
+   release process.
+
+That last point bears elaborating a bit. Unlike all system images, each delta
+image is only compatible with a single original version. In other words, while a
+system image for firmware v1.2.0 can upgrade any prior version to v1.2.0, a
+delta update for "v1.1.0→v1.2.0" can only upgrade devices running v1.1.0 to
+v1.2.0.
```
[the remark this answers is not in the document]
Comment (franc0is) [reply _drafts/ota-delta-updates.md:106]: Good call will add a note about that

File: _drafts/ota-delta-updates.md [hunk 9]
```diff
@@ -0,0 +1,607 @@
...
+> than made sense in the context of this blog post, but it may be the right
+> approach for your project. As always, your mileage may vary!
+
+## Delta Updates Implementation
+
+### Setup
+
+Like we did our previous posts, we use [Renode]({% post_url
+2020-03-23-intro-to-renode %}) to run the examples in this post. The previous
+post contains detailed instructions, but in short:
```
[the remark this answers is not in the document]
Comment (franc0is) [reply _drafts/ota-delta-updates.md:142]: ✅

**Comment by chrisc11** [comment]
late to the party but just read through this -- great post!

**Pull request merged**